### ISSUE TITLE
A Pseudo Relation

### ISSUE THREAD

**Issue by browner12** [laravel/internals#47]
here is a situation I run into with a site of mine quite often. I have `Team`s that are comprised of `Player`s. Over the course of many seasons, a `Player` can be on many `Team`s. `Coach`es add `Player`s to `Team`s. Ideally the `Player` is also a `User` on the site as well, so they can track all of the `Team`s they are on. The reality of it is, more often than not, `Player`s will not not be `User`s. 

Even if a `Player` eventually becomes a `User`, when the `Coach` adds the `Player`, they may not be a `User` yet.  Therefore, when creating their `Team`, `Coach`es fill in first_name, last_name, and email for the `Player`.  My Player table looks like this:

**Player**
- id
- team_id
- first_name
- last_name
- email
- user_id (nullable)

If someone **does** sign up, they can connect this `Player` to their `User`.

here is where the problem arises.  On my `Player` model, I have a relationship for the `User`.

``` php
public function user()
{
    return $this->belongsTo(User::class);
}
```

However, if this `Player` has not been connected to a `User`, this doesn't work. What I would like to do in this case is create a 'Pseudo User' that is returned instead.

``` php
public function user()
{
    if($this->user_id){
        return $this->belongsTo(User::class);
    }

    $user = new User();
    $user->first_name = $this->first_name;
    $user->last_name = $this->last_name;
    $user->email = $this->email;

    return $user;
}
```

The problem here is that we cannot return a `Model`, we need to return a `Relation`. I'm not exactly sure how this would work, but I would like to create our own `Relation` and populate the `getResults()` method with our Pseudo User.

Some people may ask, why not just actually create the pseudo user in the User table.  I have gone back and forth on this, and when I made the decision years ago, I decided not to. Not sure if it was the right decision, but it's what I went with.

Looking for some thoughts on this.

Thanks!

**Comment by jlaswell** [comment]
tl;dr at the end of this response

---

I'm not feeling like a pseudo relation would be a good idea in core itself, but this feels like a potential place to use the [null object pattern](https://en.wikipedia.org/wiki/Null_Object_pattern), which [Adam Watham spoke really well to](http://adamwathan.me/2015/09/02/chasing-perfect-at-laracon-eu/) at both Laracons! It may be a bit more of cramming a pattern into fitting the need, but you may be able to have a `Player` be some `User` interface that would could be implemented as a `RegisteredPlayer` or a `GuestPlayer`. A `GuestPlayer` could then have no-op functionality for that of a `RegisteredPlayer`.

Quick example,
```php
// $player is some implementation of the User interface, either Registered or Guest
$player->updateStatistics($game);

// User interface class
public function updateStatistics(Game $game): Collection;

// RegisteredPlayer
public function updateStatistics(Game $game): Collection
{
   ... // Fancy stat calculation stuffs
   return collect($updatedStats);
}

// GuestPlayer
public function updateStatistics(Game $game): Collection
{
    return collect();
}
```
For the very few times I've used this, I always find that it uncovers things like, "Maybe it would better to tackle this in x class, rather than doing this based on a relation." Of course none of this may work for your situation or be a massive headache, but hopefully this gets some thoughts going. You may actually be comfortable to implement your own [Relation](https://github.com/laravel/framework/blob/5.2/src/Illuminate/Database/Eloquent/Relations/Relation.php) class that just returns an empty array or null for `getResults()` as well. ^.^

---

tl;dr Maybe a something like the [null object pattern](https://en.wikipedia.org/wiki/Null_Object_pattern) could help here

**Comment by franzliedke** [comment]
@jlaswell nails it. :+1:

**Comment by tomschlick** [comment]
Yeah this is more of a problem about implementation than tooling. @jlaswell has the correct approach in mind

**Comment by browner12** [comment]
sorry for the **very** slow reply.

@jlaswell I absolutely agree the null object pattern is a good way to go. however, what I'd like to address is I don't think Laravel is setup for the Null pattern to work with relationships. I should have some PRs ready either today or tomorrow that I'll make sure to link here, so you guys can address.

Thanks for the feedback.

**Comment by themsaid** [comment]
Closing since relationships now can return default values. https://github.com/laravel/framework/pull/19733 for example

**Comment by browner12** [comment]
yup, good by me.

defaults is the way we decided to go.
